Sources/NIOSSL/SSLCertificateExtensions.swift

```diff
@@ -0,0 +1,221 @@
+            self = .borrowed(reference: reference, owner: owner)
+        }
+        
+        /// The owner of the memory to which the reference points
+        var owner: AnyObject {
+            switch self {
+            case .owned(let deallocator):
+                return deallocator
+            case .borrowed(_, let owner):
+                return owner
+            }
+        }
+        
+        /// All operations accessing `reference` need to be implemented while guaranteeing that we still have a reference to the memory owner.
+        /// Otherwise `reference` could already be freed. This would result in undefined behaviour as we access a dangling pointer.
+        /// This method guarantees that `reference` is valid during execution of `body`.
+        internal func withReference<Result>(
+            _ body: (OpaquePointer?) throws -> Result
+        ) rethrows -> Result {
+            try withExtendedLifetime(self) {
+                switch self {
+                case .owned(let deallocator):
+                    return try body(deallocator.reference)
+                case .borrowed(let reference, _):
+                    return try body(reference)
+                }
+            }
+        }
+    }
+    
+    @usableFromInline internal let stackSize: Int
+    private let storage: Storage
+    
+    internal init(takeOwnershipOf reference: OpaquePointer?) {
+        self.storage = .init(takeOwnershipOf: reference)
+        if let reference = reference {
+            self.stackSize = CNIOBoringSSL_sk_X509_EXTENSION_num(reference)
+        } else {
+            self.stackSize = 0
+        }
+    }
+    
+    internal init(borrowing reference: OpaquePointer?, owner: AnyObject) {
+        self.storage = .init(borrowing: reference, owner: owner)
+        if let reference = reference {
+            self.stackSize = CNIOBoringSSL_sk_X509_EXTENSION_num(reference)
+        } else {
+            self.stackSize = 0
+        }
+    }
+    
+    internal init(copyOf reference: OpaquePointer?) {
+        guard let reference = reference else {
+            self.init()
+            return
+        }
+        self.init(takeOwnershipOf: CNIOBoringSSL_sk_X509_EXTENSION_dup(reference))
+    }
+    
+    internal init() {
+        self.init(takeOwnershipOf: nil)
+    }
+}
+
+extension NIOSSLCertificate {
+    public var _extensions: _NIOSSLCertificateExtensions {
+        _NIOSSLCertificateExtensions(borrowing: CNIOBoringSSL_X509_get0_extensions(self._ref), owner: self)
+    }
+}
+
+extension _NIOSSLCertificateExtensions: RandomAccessCollection {
+    public subscript(position: Int) -> _NIOSSLCertificateExtension {
+        precondition(self.indices.contains(position), "index \(position) out of bounds")
+        return self.storage.withReference { reference in
+            guard let reference = reference else {
+                fatalError("unexpected null pointer when unwrapping reference for value at \(position)")
```

Comment:
This is just a fancy force-unwrap, so we can use a force-unwrap and save the code.